homeassistant/components/http.py

```diff
@@ -106,413 +68,347 @@ def setup(hass, config):
+        """Register a folder to serve as a static path."""
+        from static import Cling
 
-    def set_cache_header(self):
-        """Add cache headers if not in development."""
-        if self.server.development:
-            return
+        headers = []
 
-        # 1 year in seconds
-        cache_time = 365 * 86400
+        if not self.development:
+            # 1 year in seconds
+            cache_time = 365 * 86400
 
-        self.send_header(
-            HTTP_HEADER_CACHE_CONTROL,
-            "public, max-age={}".format(cache_time))
-        self.send_header(
-            HTTP_HEADER_EXPIRES,
-            self.date_time_string(time.time()+cache_time))
+            headers.append({
+                'prefix': '',
+                HTTP_HEADER_CACHE_CONTROL:
+                "public, max-age={}".format(cache_time)
+            })
 
-    def set_session_cookie_header(self):
-        """Add the header for the session cookie and return session ID."""
-        if not self.authenticated:
-            return None
+        self.register_wsgi_app(url_root, Cling(path, headers=headers))
 
-        session_id = self.get_cookie_session_id()
+    def register_wsgi_app(self, url_root, app):
+        """Register a path to serve a WSGI app."""
+        if url_root in self.extra_apps:
+            _LOGGER.warning("Url root '%s' is being overwritten", url_root)
 
-        if session_id is not None:
-            self.server.sessions.extend_validation(session_id)
-            return session_id
+        self.extra_apps[url_root] = app
 
-        self.send_header(
-            'Set-Cookie',
-            '{}={}'.format(SESSION_KEY, self.server.sessions.create())
+    def start(self):
+        """Start the wsgi server."""
+        from eventlet import wsgi
+        import eventlet
+
+        sock = eventlet.listen((self.server_host, self.server_port))
+        if self.ssl_certificate:
+            eventlet.wrap_ssl(sock, certfile=self.ssl_certificate,
+                              keyfile=self.ssl_key, server_side=True)
+        wsgi.server(sock, self)
+
+    def dispatch_request(self, request):
+        """Handle incoming request."""
+        from werkzeug.exceptions import (
+            MethodNotAllowed, NotFound, BadRequest, Unauthorized,
         )
+        from werkzeug.routing import RequestRedirect
 
-        return session_id
-
-    def verify_session(self):
-        """Verify that we are in a valid session."""
-        return self.get_cookie_session_id() is not None
+        with request:
+            adapter = self.url_map.bind_to_environ(request.environ)
+            try:
+                endpoint, values = adapter.match()
+                resp = self.views[endpoint].handle_request(request, **values)
+
+                if resp.is_streamed or 200 > resp.status_code >= 300 or \
+                   'gzip' not in request.accept_encodings:
+                    return resp
+
+                resp.headers.extend([('Content-Encoding', 'gzip'),
```

Comment:
We might want to restrict the things that we gzip. For already compressed formats (images) or small files it can make the size bigger. I am not sure if the best approach would be to just limit it to text responses?

CC @JshWright